File: 网络编程/Python/tcp_server.py
#!usr/bin/env python3
# -*- coding: utf-8 -*-

"""
TCP server demo.
"""

__author__ = 'Ziang Lu'

# import concurrent.futures as cf
import socket
import socketserver
import time
from typing import Tuple


def tcp_worker(sock_conn: socket.socket, addr: Tuple[str, int]) -> None:
    """
    Thread function to handle TCP connection.
    :param sock_conn: socket
    :param addr: tuple(str, int)
    :return: None
    """
    host, port = addr
    print(f'[SERVER] Connection accepted from {host}:{port}')
    sock_conn.send(b'Welcome!')
    while True:
        data = sock_conn.recv(1024)
        time.sleep(1)
        if not data or data.decode('utf-8') == 'exit':
            break
        sock_conn.send(f"Hello, {data.decode('utf-8')}".encode('utf-8'))
    sock_conn.close()
    print(f'[SERVER] Connection accepted from {host}:{port} CLOSED')


##### METHOD 1: With "socket" module #####


# # Create an IPv4, TCP socket
# server_sock = socket.socket(family=socket.AF_INET, type=socket.SOCK_STREAM)
# # Bind the socket to server address "127.0.0.1:9999"
# server_sock.bind(('127.0.0.1', 9999))
# print('[SERVER] Server bound to 127.0.0.1:9999')

# # Let the server start listening for connections requests
# server_sock.listen()  # Becomes a server socket
# print('[SERVER] Listening for connection...')

# # => Use a thread pool to reuse the thread, and thus improve performance
# with cf.ThreadPoolExecutor(max_workers=50) as pool:
#     try:
#         while True:
#             sock_conn, addr = server_sock.accept()  # Accepted a connection
#             # We want to fire up a thread to handle the connection, so that the
#             # server is not blocked away from other connections.
#             # => Use a thread pool to reuse the thread, and thus improve
#             #    performance
#             pool.submit(tcp_worker, sock_conn, addr)
#     except KeyboardInterrupt:
#         # Close the server socket
#         server_sock.close()


##### METHOD 2: With "socketserver" module #####


class MyTCPHandler(socketserver.BaseRequestHandler):
    """
    TCP handler class for our TCP server.
    When a connection is accepted, a thread is fired up to handle it, an object
    of this class is instantiated, and the corresponding handle() method is
    called.
    """

    def handle(self):
        sock_conn = self.request
        # Reuse the "thread" function defined above
        tcp_worker(sock_conn, self.client_address)


# Create a multi-threaded version of TCP server, and bind the server socket to
# server address "127.0.0.1:9999"
server = socketserver.ThreadingTCPServer(
    server_address=('127.0.0.1', 9999), RequestHandlerClass=MyTCPHandler
)
# With ThreadingTCPServer, each connection will have a thread fired up to handle
# it, a RequestHandlerClass object is instantiated, and the corresponding
# handle() method is called.
print('[SERVER] Server bound to 127.0.0.1:9999')

try:
    # Activate the server
    # Let the server start listening for connections requests
    print('[SERVER] Listening for connections...')
    server.serve_forever()
except KeyboardInterrupt:
    # Close the server
    server.server_close()

# Output:
# [SERVER] Server bound to 127.0.0.1:9999
# [SERVER] Listening for connection...
# [SERVER] Connection accepted from 127.0.0.1:56389
# [SERVER] Connection accepted from 127.0.0.1:56389 CLOSED

File: 网络编程/Java/TCPServer.java
import java.io.IOException;
import java.io.DataInputStream;
import java.io.DataOutputStream;
import java.net.InetSocketAddress;
import java.net.ServerSocket;
import java.net.Socket;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

/**
 * TCP server demo.
 *
 * @author Ziang Lu
 */
public class TCPServer {

    /**
     * Thread function to handle TCP connection.
     */
    private static class TCPWorker implements Runnable {
        /**
         * TCP socket connection.
         */
        private Socket sock;

        /**
         * Constructor with parameter.
         * @param sock TCP socket connection
         */
        TCPWorker(Socket sock) {
            this.sock = sock;
        }

        @Override
        public void run() {
            System.out.println(String.format("[SERVER] Connection accepted from %s", sock.getRemoteSocketAddress()));
            try {
                DataOutputStream out = new DataOutputStream(sock.getOutputStream());
                DataInputStream in = new DataInputStream(sock.getInputStream());

                out.writeUTF("Welcome!");
                out.flush();
                while (true) {
                    String data = in.readUTF();
                    Thread.sleep(1000);
                    if (data.equals("") || data.equals("exit")) {
                        break;
                    }
                    out.writeUTF("Hello, " + data);
                    out.flush();
                }
                sock.close();
            } catch (IOException | InterruptedException e) {
                e.printStackTrace();
            }
            System.out.println(
                    String.format("[SERVER] Connection accepted from %s CLOSED", sock.getRemoteSocketAddress()));
        }
    }

    /**
     * Main driver.
     * @param args arguments from command line
     */
    public static void main(String[] args) {
        try {
            // Create an IPv4, TCP server socket
            ServerSocket serverSocket = new ServerSocket();
            // Bind the server socket to server address "127.0.0.1:9999", and let the server start listening for
            // connections requests
            serverSocket.bind(new InetSocketAddress("127.0.0.1", 9999));
            System.out.println("[Server] Server bound to 127.0.0.1:9999");
            System.out.println("[SERVER] Listening for connection...");

            // => Use a thread pool to reuse the thread, and thus improve performance
            ExecutorService pool = Executors.newFixedThreadPool(50);
            while (true) {
                Socket sock = serverSocket.accept(); // Accepted a connection
                // We want to fire up a thread to handle the connection, so that the server is not blocked away from
                // other connections.
                // => Use a thread pool to reuse the thread, and thus improve performance
                pool.submit(new TCPWorker(sock));
            }
        } catch (IOException e) {
            e.printStackTrace();
        }

        /*
         * Output:
         * [SERVER] Server bound to 127.0.0.1:9999
         * [SERVER] Listening for connection...
         * [SERVER] Connection accepted from /127.0.0.1:59808
         * [SERVER] Connection accepted from /127.0.0.1:59808 CLOSED
         */
    }

}
